### PR TITLE
modify create_time_series to save RESTOM time series file

### PR DESCRIPTION
**Pull request by brianpm**
Save RESTOM time series file whenever FLNT and FSNT are available  in the `hist_file_var_list` in `create_time_series`. 

This is a little brute force, but is relatively low overhead, and then allows RESTOM to be used like a regular variable throughout the rest of ADF.

### CODE REVIEW

**Comment by justin-richling** [comment]
I forgot that the change in `amwg_table.py` assumes we add RESTOM to the diag_var_list in the config files

**Comment by brianpm** [comment]
@justin-richling --- I hope I did this correctly. I "synced" my branch with ADF main, and then I put your `amwg_table.py` into a commit and pushed back to my branch. Looks like one of the tests is failing, but I don't know why.

**Comment by nusbaume** [comment]
Just wanted to add a note here to see if we can add RESTOM to the time series files via the same method used to add PRECT in PR #260?  If you need any help or guidance doing that just let me know (and once that is done I'll go ahead and do a formal review).  Thanks!

**Comment by brianpm** [comment]
I have modified `adf_diag.py` to derive `RESTOM` following the same recipe/approach as #260. I tested that RESTOM time series files are being generated with real CAM data. I also added `RESTOM` to `adf_variable_defaults.yaml` because the code does not know it is a derivable quantity otherwise (Note: I did not check if the colors/contour levels are good, and that will probably need to be adjusted.)

A couple of things I noted:
- In both cases (`PRECT` and `RESTOM`) this approach assumes that there will be one time series file per variable; the method may or may not succeed in creating a new file in other situations, but in all other possible situations, the result file will not be the correct/complete result.
- The above problem also includes finding ".tmp" files from previous failed attempts.
- There was no provision for what to do in cases where the derived variable file already exists, which would probably end up with an NCO prompt at the command line. I added a simple check and a kwarg to determine whether to overwrite or not. If overwrite is true and the file exists, it will be removed and re-made, otherwise it will use the existing file. Applied this to both `PRECT` and `RESTOM`.
- this approach produces files that will have all three variables in them, replicating the time series information from the two input variables
- this approach does not correct metadata for the derived variable, it is copied from one of the input variables. This may negatively impact automatic labeling of plots (especially long_name is incorrect).  
- The use of `os.system` has been superseded by the [subprocess approach](https://docs.python.org/3/library/subprocess.html)
- The indentation of the code was weird, so I reformatted using the [black](https://github.com/psf/black) formatter (I tried to do just the method at the end, but I think it hit the whole file).

File: lib/adf_diag.py [hunk 1]
```diff
@@ -778,185 +810,273 @@ def create_plots(self):
...
+                # create new file name for RESTOM
+                derived_file = constit_files[0].replace("FLNT", "RESTOM")
+                if Path(derived_file).is_file():
+                    if overwrite:
+                        Path(derived_file).unlink()
+                    else:
+                        print(
+                            f"[{__name__}] Warning: RESTOM file was found and overwrite is False. Will use existing file."
+                        )
+                        return None
```
Comment (nusbaume) [inline lib/adf_diag.py:1076]: I think you want to use `continue` here instead:

```suggestion
                        continue
```

**Review by nusbaume — APPROVED** [review]
Looks good!  I did have a couple of suggested changes and questions, but nothing that should meaningfully hold up this PR, at least on my end.

Also, I tried to capture your concerns (along with at least one of my own) with the current derived variables method with ADF issue #273.  Please feel free to add anything I missed!

File: scripts/analysis/amwg_table.py [hunk 2]
```diff
@@ -305,35 +291,15 @@ def amwg_table(adf):
...
-            #Print message to debug log:
-            adf.debug_log("RESTOM not calculated because FSNT and/or FLNT variables not in dataset")
+        # Move RESTOM to top of table
+        #----------------------------
+        if 'RESTOM' in var_list:
+            table_df = pd.read_csv(output_csv_file)
+            idx = table_df.index[table_df['variable'] == 'RESTOM'].tolist()[0]
+            table_df = pd.concat([table_df[table_df['variable'] == 'RESTOM'], table_df]).reset_index(drop = True)
+            table_df = table_df.drop([idx+1]).reset_index(drop=True)
+            table_df = table_df.drop_duplicates()
```
[the remark this answers is not in the document]
Comment (justin-richling) [reply scripts/analysis/amwg_table.py:301]: Just tested it, yeah the  lines 
```python
idx = table_df.index[table_df['variable'] == 'RESTOM'].tolist()[0]
```
and
```python
table_df = table_df.drop([idx+1]).reset_index(drop=True)
```
are not needed.

**Review by justin-richling — COMMENTED** [review]
This looks good to me, my suggestions can be done here, or in a different PR if desired.
Thanks @brianpm

**Review by justin-richling — APPROVED** [review]

File: scripts/analysis/amwg_table.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (brianpm) [reply scripts/analysis/amwg_table.py:301]: Thanks. These lines are removed.

File: lib/adf_diag.py [hunk 3]
```diff
@@ -778,185 +810,273 @@ def create_plots(self):
...
+                # create new file name for PRECT
+                prect_file = constit_files[0].replace("PRECC", "PRECT")
+                if Path(prect_file).is_file():
+                    if overwrite:
+                        Path(prect_file).unlink()
+                    else:
+                        print(
+                            f"[{__name__}] Warning: PRECT file was found and overwrite is False. Will use existing file."
+                        )
+                        return None
```
[the remark this answers is not in the document]
Comment (brianpm) [reply lib/adf_diag.py:1043]: You're right. I changed it.

File: lib/adf_diag.py [hunk 4]
```diff
@@ -778,185 +810,273 @@ def create_plots(self):
...
+                    if overwrite:
+                        Path(derived_file).unlink()
+                    else:
+                        print(
+                            f"[{__name__}] Warning: RESTOM file was found and overwrite is False. Will use existing file."
+                        )
+                        return None
+                # append FSNT to the file containing FLNT
+                os.system(f"ncks -A -v FLNT {constit_files[0]} {constit_files[1]}")
+                # create new file with the sum of FLNT and FSNT
```
[the remark this answers is not in the document]
Comment (brianpm) [reply lib/adf_diag.py:1079]: Done.

File: lib/adf_diag.py (hunk 1 above)
Comment (brianpm) [reply lib/adf_diag.py:1076]: Done.

File: lib/adf_diag.py [hunk 5]
```diff
@@ -585,6 +590,31 @@ def call_ncrcat(cmd):
...
+                shortwavenet = ts_dir[case_idx] + os.sep + \
+                ".".join([case_name, hist_str, swnet_name, time_string, "nc" ])
+                restom_fil = ts_dir[case_idx] + os.sep + \
+                ".".join([case_name, hist_str, "RESTOM", time_string, "nc" ])
+                # combine into a single file
+                tmp1 = xr.open_dataset(longwavenet)
+                tmp2 = xr.open_dataset(shortwavenet)
+                tmp3 = tmp2[swnet_name]-tmp1[lwnet_name]
+                tmp3.name = "RESTOM"
+                tmp3.attrs['long_name'] = "residual radiative flux at top of model"
```
[the remark this answers is not in the document]
Comment (brianpm) [reply lib/adf_diag.py]: These lines are gone now.

File: lib/adf_variable_defaults.yaml [hunk 6]
```diff
@@ -683,6 +683,19 @@ QRS:
 #+++++++++++++++++
 # Category: TOA energy flux
 #+++++++++++++++++
+RESTOM:
+  colormap: "RdBu_r"
+  contour_levels_range: [-10, 10, 0.5]
```
[the remark this answers is not in the document]
Comment (brianpm) [reply lib/adf_variable_defaults.yaml:688]: Yeah, totally a typo. I just made the range from -100 to 100 in the updated code.

**Comment by brianpm** [comment]
Okie dokie, I think I just addressed all the issues from the review.

**Review by justin-richling — APPROVED** [review]
Looks good @brianpm, thanks for getting this in!

**Pull request merged**